TestProjects/UniversalGraphicsTest_Lighting/Assets/Scenes/146_Unlit_Custom_Shadows/CustomLighting.hlsl

```diff
@@ -0,0 +1,86 @@
+//#pragma multi_compile _ _MAIN_LIGHT_SHADOWS _MAIN_LIGHT_SHADOWS_CASCADE _MAIN_LIGHT_SHADOWS_SCREEN
+
+void MainLight_half(float3 WorldPos, out half3 Direction, out half3 Color, out half DistanceAtten, out half ShadowAtten)
+{
+#ifdef SHADERGRAPH_PREVIEW
+   Direction = half3(0.5, 0.5, 0);
+   Color = 1;
+   DistanceAtten = 1;
+   ShadowAtten = 1;
+#else
+   half4 shadowCoord = TransformWorldToShadowCoord(WorldPos);
+   Light mainLight = GetMainLight(shadowCoord);
+   Direction = mainLight.direction;
+   Color = mainLight.color;
+   DistanceAtten = mainLight.distanceAttenuation;
+   ShadowAtten = mainLight.shadowAttenuation;
+#endif
+}
+
+void MainLight_half(out half3 Direction, out half3 Color, out half DistanceAtten, out half ShadowAtten)
+{
+#ifdef SHADERGRAPH_PREVIEW
+   Direction = half3(0.5, 0.5, 0);
+   Color = 1;
+   DistanceAtten = 1;
+   ShadowAtten = 1;
+#else
+   Light mainLight = GetMainLight();
+   Direction = mainLight.direction;
+   Color = mainLight.color;
+   DistanceAtten = mainLight.distanceAttenuation;
+   ShadowAtten = mainLight.shadowAttenuation;
+#endif
+}
+
+void MainLightFull_half(float3 WorldPos, out half3 Direction, out half3 Color, out half DistanceAtten, out half ShadowAtten)
+{
+   #ifdef SHADERGRAPH_PREVIEW
+   Direction = half3(0.5, 0.5, 0);
+   Color = 1;
+   DistanceAtten = 1;
+   ShadowAtten = 1;
+   #else
+   #ifdef SHADOWS_SCREEN
+   half4 clipPos = TransformWorldToHClip(WorldPos);
+   half4 shadowCoord = ComputeScreenPos(clipPos);
+   #else
+   half4 shadowCoord = TransformWorldToShadowCoord(WorldPos);
+   #endif
+   Light mainLight = GetMainLight(shadowCoord);
+   Direction = mainLight.direction;
+   Color = mainLight.color;
+   DistanceAtten = mainLight.distanceAttenuation;
+   ShadowAtten = mainLight.shadowAttenuation;
+   #endif
+}
+
+void SampleSH_half(half3 normalWS, out half3 Ambient)
+{
+    // LPPV is not supported in Ligthweight Pipeline
+    real4 SHCoefficients[7];
```

Comment:
Can you move these assignments into an array initializer list so that it is just one line?

Reply:
It is brutal how in the function, it just splits them out as well. Having too many parameters can impact compilation and fitting things in registers, but I can't see this as being better. 
float3 SampleSH9(float4 SHCoefficients[7], float3 N)
{
    float4 shAr = SHCoefficients[0];
    float4 shAg = SHCoefficients[1];
    float4 shAb = SHCoefficients[2];
    float4 shBr = SHCoefficients[3];
    float4 shBg = SHCoefficients[4];
    float4 shBb = SHCoefficients[5];
    float4 shCr = SHCoefficients[6];

Reply:
Do you know if there is a reason it is done this way? Or can we just modify the helper function to remove the temporaries?

Reply:
IDK this is from our URP as these a test function. I think GPU will eleminate these temporaries any way